src/gl/context.cpp

```diff
@@ -522,101 +519,54 @@ void Context::submit_pssa(const unsigned surface_index, mat4x4 sun_view) {
   pixel_areas.at(surface_buffer.index) = pixel_area;
 }
 
-void Context::submit_pssa(const std::vector<unsigned> &surface_indices, mat4x4 sun_view) {
+void Context::submit_pssa(const unsigned int surface_index, mat4x4 sun_view) {
+  submit_pssa(model.surface_buffers[surface_index], sun_view);
+}
+
+void Context::submit_pssas(const std::vector<unsigned int> &surface_indices, mat4x4 sun_view) {
   for (auto const surface_index : surface_indices) {
-    auto const &surface_buffer = model.surface_buffers[surface_index];
-    auto const pixel_area = set_scene(sun_view, &surface_buffer);
-    draw_model();
-    glBeginQuery(GL_SAMPLES_PASSED, queries.at(surface_buffer.index));
-    GLModel::draw_surface(surface_buffer);
-    glEndQuery(GL_SAMPLES_PASSED);
-    pixel_areas.at(surface_buffer.index) = pixel_area;
+    submit_pssa(surface_index, sun_view);
   }
 }
 
 void Context::submit_pssa(mat4x4 sun_view) {
   for (auto const &surface_buffer : model.surface_buffers) {
-    auto const pixel_area = set_scene(sun_view, &surface_buffer);
-    draw_model();
-    glBeginQuery(GL_SAMPLES_PASSED, queries.at(surface_buffer.index));
-    GLModel::draw_surface(surface_buffer);
-    glEndQuery(GL_SAMPLES_PASSED);
-    pixel_areas.at(surface_buffer.index) = pixel_area;
+    submit_pssa(surface_buffer, sun_view);
   }
 }
 
-float Context::calculate_pssa(const unsigned surface_index) {
-
-  if (is_render_mode) { // if currently render mode, switch to off-screen mode
-    initialize_off_screen_mode();
-    is_render_mode = false;
-  }
-
-  // wait until the result is available
-  GLint ready(0);
-  while (!ready) {
-    glGetQueryObjectiv(queries[surface_index], GL_QUERY_RESULT_AVAILABLE, &ready);
-  }
-
-  // retrieve result
+float Context::retrieve_pssa(const unsigned int surface_index) {
   glGetQueryObjectiv(queries[surface_index], GL_QUERY_RESULT, &(pixel_counts.at(surface_index)));
-
   return static_cast<float>(pixel_counts[surface_index]) * pixel_areas[surface_index];
 }
 
-std::vector<float> Context::calculate_pssa(const std::vector<unsigned> &surface_indices) {
-
-  if (is_render_mode) { // if currently render mode, switch to off-screen mode
-    initialize_off_screen_mode();
-    is_render_mode = false;
-  }
-
-  // retrieve result
-  std::vector<float> results;
-  results.reserve(surface_indices.size());
-  for (auto const &surface_index : surface_indices) {
-    glGetQueryObjectiv(queries[surface_index], GL_QUERY_RESULT, &(pixel_counts.at(surface_index)));
-    results.emplace_back(static_cast<float>(pixel_counts[surface_index]) *
-                         pixel_areas[surface_index]);
+std::vector<float> Context::retrieve_pssas(const std::vector<unsigned int> &surface_indices) {
+  std::vector<float> pssas;
+  pssas.reserve(surface_indices.size());
+  for (const unsigned int surface_index : surface_indices) {
+    pssas.emplace_back(retrieve_pssa(surface_index));
```

Comment:
Nothing to change here, but just a note that push_back is as good or better than emplace_back in most cases, so we should just be a bit circumspect about replacing it unless we need to forward arguments or use an explicit constructor.

Reply:
I was wondering about this, especially in the context of `std::vector<float>`. I think it's entirely different when it's a vector of complex objects where you don't want to construct and copy.

Reply:
Agreed. But I don't see a downside, here anyway.